### PR TITLE
Piccoloman/client api

### PR DESCRIPTION
**Pull request by piccoloman**
Ready for testing

### CODE REVIEW

File: src/crypto.ts [hunk 1]
```diff
@@ -0,0 +1,338 @@
+// TODO figure better way to share with lib crypto on tiplink site
+// TODO ts-ignore's for cryptoKey issues?
+import type { Keypair } from "@solana/web3.js";
+import nacl from "tweetnacl";
```
Comment (wisefool769) [inline src/crypto.ts:4]: It's a little bit unfortunate that we have nacl, crypto, AND libsodium to do crypto operations in here.

File: src/crypto.ts [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+  return {
+    nonceUint8Arr: nonce,
+    nonceBs58: bs58.encode(nonce),
+    encryptedDataUint8Arr: encryptedData,
+    encryptedDataBs58: bs58.encode(encryptedData),
+    encryptionKeypairToHold: keypair,
+  };
+};
+
+export const decryptDataUsingDH = (
```
Comment (wisefool769) [inline src/crypto.ts:308]: I assume this stuff is essentially copy-paste from the pro side right? so the entire crypto class is not really public-facing.

File: package.json [hunk 3]
```diff
@@ -17,7 +17,11 @@
     "@solana/web3.js": "^1.48.0",
     "bs58": "^5.0.0",
     "libsodium": "^0.7.11",
-    "libsodium-wrappers-sumo": "^0.7.11"
+    "libsodium-wrappers-sumo": "^0.7.11",
+    "nanoid": "^3.3.6",
+    "sodium-plus": "^0.9.0",
```
Comment (wisefool769) [inline package.json:22]: We need to add back sodium-plus o.O

File: src/client.ts [hunk 4]
```diff
@@ -0,0 +1,538 @@
...
+    this.apiKey = apiKey;
+    this.version = version;
+    this.campaigns = new CampaignActions({client: this});
+  }
+
+  public static async init(apiKey: string, version: number = 1): Promise<TipLinkClient> {
+    const client = new TipLinkClient(apiKey, version);
+
+    const apiKeyRes = await client.fetch("api_key");
+    client.id = apiKeyRes['account']['id'];
```
Comment (wisefool769) [inline src/client.ts:37]: Should apiKeyRes have a type?

File: src/client.ts [hunk 5]
```diff
@@ -0,0 +1,538 @@
...
+    client.publicKey = apiKeyRes['account']['public_key'];
+
+    return client;
+  }
+
+  // TODO type return?
+  public async fetch(endpoint: string, args: Record<string, any> | null = null, body: Record<string, any> | null = null, verb: string="GET"): Promise<any> {
```
Comment (wisefool769) [inline src/client.ts:44]: Do we need a public-facing fetch function?

File: src/client.ts [hunk 6]
```diff
@@ -0,0 +1,538 @@
...
+  description: string;
+  imageUrl: string;
+  active: boolean; // default true
+
+  limit: number;
+  offset: number;
+  sorting: string;
+}
+
+class DispenserActions extends TipLinkApi {
```
Comment (wisefool769) [inline src/client.ts:377]: This API is kind of interesting. I guess I would have expected there to just be a crapton of calls nested under TIpLinkClient? Maybe this is better though. 

I don't know why we want to expose the actual underlying client though bc people don't know what endpoints to hit?

File: src/client.ts [hunk 7]
```diff
@@ -0,0 +1,538 @@
...
+
+export class TipLinkClient {
+  apiKey: string;
+  version: number;
+
+  // Set from apikey
+  publicKey?: string;
+  id?: number;
+
+  campaigns: CampaignActions;
```
Comment (wisefool769) [inline src/client.ts:25]: ohhhhh i understand now, sick

File: src/client.ts [hunk 8]
```diff
@@ -0,0 +1,538 @@
...
+        campaign_id: campaign.id,
+        account_id: this.client.id,
+        is_owner: true,
+        encrypted_campaign_key: encryptedCampaignKey,
+      };
+
+      const joinResult = await this.client.fetch(`campaigns/${campaign.id}/campaign_account_joins`, null, keyData, "POST");
+
+      if(salt !== res['encryption_salt']) {
+        console.error("encryption salt mismatch");
```
Comment (wisefool769) [inline src/client.ts:162]: when does this happen?

File: src/client.ts [hunk 9]
```diff
@@ -0,0 +1,538 @@
...
+
+    const res = (await this.client.fetch("campaigns", findParams, null, "GET") as Record<string, any>[])[0];
+
+    const campaign = new Campaign({client: this.client, id: res['id'], name: res['name'], description: res['description'], imageUrl: res['image_url'], active: res['active'],});
+
+    campaign.encryptionSalt = res['encryption_salt'];
+
+    console.warn("unable to acquire decryption key");
+    // const encryptedKeyRes = await this.client.fetch(`campaigns/${campaign.id}/campaign_account_joins`);
+    // // TODO figure out how api keys will do key exchange / sharing to get privateKey
```
Comment (wisefool769) [inline src/client.ts:187]: What are your current thoughts on how to do this bit?

File: src/client.ts [hunk 10]
```diff
@@ -0,0 +1,538 @@
...
+  dispensers: DispenserActions;
+
+  public constructor(
+    params: CampaignConstructorParams
+    // client: TipLinkClient, id: number, name: string, description: string, imageUrl: string, active: boolean = true
+  ) {
+    super(params.client);
+    this.id = params.id;
+    this.name = params.name;
+    this.description = params.description;
```
Comment (wisefool769) [inline src/client.ts:263]: I guess I don't love keeping state for all these things? They can just get out of sync right? 

I like having an interface for the Dispenser and  some params for construction, but do we need to keep track of this?

File: src/crypto.ts (hunk 2 above)
Comment (piccoloman) [reply src/crypto.ts:308]: Yeah, we won't doc this or anything

File: src/client.ts (hunk 4 above)
Comment (piccoloman) [reply src/client.ts:37]: Yeah, we can probably type it, I am not exactly sure how we want to handle types / cross repo things. Like should we just copy paste like the crypto stuff. Should we do importing from one to the other.

File: src/client.ts (hunk 5 above)
Comment (piccoloman) [reply src/client.ts:44]: Public is necessary for other class to use it (like campaigns etc). I'm not planning to publicize it i.e. include it in the docs.

File: src/client.ts (hunk 8 above)
Comment (piccoloman) [reply src/client.ts:162]: I think it shouldn't ever happen. Might be possible with some weird race condition on the server, or if some db constraint fails to save it or something.

File: src/client.ts (hunk 9 above)
Comment (piccoloman) [reply src/client.ts:187]: We should talk this one over a call I think

File: src/client.ts (hunk 10 above)
Comment (piccoloman) [reply src/client.ts:263]: We don't need to in some sense, but then if they ever want to get the information we'll have to write getter functions and make a fetch call to the server.

File: src/client.ts (hunk 4 above)
Comment (wisefool769) [reply src/client.ts:37]: We should do importing from tiplink-api tbh

File: src/client.ts (hunk 8 above)
Comment (wisefool769) [reply src/client.ts:162]: But you have seen it?

**Review by wisefool769 — APPROVED** [review]

**Review by wisefool769 — APPROVED** [review]

**Pull request merged**